packages/flutter/lib/src/widgets/framework.dart

```diff
@@ -298,67 +398,166 @@ abstract class Widget {
+  /// widget returned by this function, either by updating the existing subtree
+  /// or by removing the subtree and inflating a new subtree, depending on
+  /// whether the widget returned by this function [Widget.canUpdate] the root
+  /// of the existing subtree.
+  ///
+  /// Typically implementations return a newly created constellation of widget
+  /// that are configured with information from this widget's constructor and
+  /// from the given [BuildContext].
+  ///
+  /// The given [BuildContext] contains information about the location in the
+  /// tree at which this widget is being built. For example, the context
+  /// provides the set of inherited widgets for this location in the tree. A
+  /// given widget might be with multiple different [BuildContext] arguments
+  /// over time if the widget is moved around the tree or if the widget is
+  /// inserted into the tree in multiple places at once.
   @protected
   Widget build(BuildContext context);
 
   /// Trampoline to make the [build] closure library-accessible.
   WidgetBuilder get _build => build;
 }
 
-/// StatefulWidgets provide the configuration for
-/// [StatefulElement]s, which wrap [State]s, which hold mutable state
-/// and can dynamically and spontaneously ask to be rebuilt.
+/// A widget that has mutable state.
+///
+/// A stateful widget is a widget that describes part of the user interface by
+/// building a constellation of other widgets that describe the user interface
+/// more concretely. The building process continues recursively until the
+/// description of the user interface is fully concreate (e.g., consists
+/// enitrely of [RenderObjectWidget]s, which describe concrete [RenderObject]s).
+///
+/// Stateless widget are useful when the part of the user interface you are
+/// describing can change dynamically, e.g. due to having an internal
+/// clock-driven state, or depending on some system state. For compositions that
+/// depend only on the configuration information in the object itself and the
+/// [BuildContext] in which the widget is inflated, consider using
+/// [StatelessWidget].
+///
+/// [StatefulWidget] instances themselves are immutable and store their mutable
+/// state in separate [State] objects that are created by the [createState]
+/// function. The framework calls [createState] whenever it inflates a
+/// [StatefulWidget], which means that multiple [State] objects might be
+/// associated with the same [StatefulWidget] if that widget has been inserted
+/// into the tree in multiple places. Similarly, if a [StatefulWidget] is
+/// removed from the tree and later inserted in to the tree again, the framework
+/// will call [createState] again to create a fresh [State] object, simplifying
+/// the lifecycle of [State] objects.
+///
+/// A [StatefulWidget] can keep the same [State] object when moving from one
+/// location in the tree to another by using a [GlobalKey] for its [key].
```

Comment:
I'd rather we suggested that people who create the Widget are able to use a GlobalKey rather that suggest that the widget itself can use it.